components/validator/challenger.go

```diff
@@ -65,191 +93,413 @@ func NewChallenger(ctx context.Context, cfg Config, l log.Logger) (*Challenger,
 	}
 
 	return &Challenger{
-		log: l,
-		cfg: cfg,
-		ctx: ctx,
+		log:  l,
+		cfg:  cfg,
+		ctx:  ctx,
+		done: make(chan struct{}),
 
 		callOpts: utils.NewCallOptsWithSender(ctx, cfg.TxManager.From()),
 		txOpts:   utils.NewSimpleTxOpts(ctx, cfg.TxManager.From(), cfg.TxManager.Signer),
 
+		l1Client: cfg.L1Client,
+
 		l2ooContract:      l2ooContract,
+		l2ooABI:           l2ooABI,
 		colosseumContract: colosseumContract,
+		colosseumABI:      colosseumABI,
 
 		submissionInterval:        submissionInterval,
 		finalizationPeriodSeconds: finalizationPeriodSeconds,
 		l2BlockTime:               l2BlockTime,
-	}, nil
-}
 
-func (c *Challenger) IsChallengeInProgress() (bool, error) {
-	return c.colosseumContract.IsInProgress(c.callOpts)
+		txCandidatesChan: txCandidateChan,
+	}, nil
 }
 
-func (c *Challenger) GetChallengeInProgress() (bindings.TypesChallenge, error) {
-	return c.colosseumContract.GetChallengeInProgress(c.callOpts)
-}
+func (c *Challenger) Start() error {
+	c.log.Info("start challenger")
 
-func (c *Challenger) OutputAtBlockSafe(blockNumber uint64, includeNextBlock bool) (*eth.OutputResponse, error) {
-	ctx, cancel := context.WithTimeout(c.ctx, c.cfg.NetworkTimeout)
-	defer cancel()
-	output, err := c.cfg.RollupClient.OutputAtBlock(ctx, blockNumber, includeNextBlock)
+	// set the challenger's checkpoint to the latest output index
+	// because when challenger starts, it is guaranteed that the past outputs are checked through scanPrevOutputs
+	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(c.callOpts)
 	if err != nil {
-		return nil, err
+		return err
 	}
+	c.checkpoint = new(big.Int).Sub(nextOutputIndex, common.Big1)
 
-	return output, nil
+	// scan all the outputs within the finalization window
+	c.wg.Add(1)
+	go c.scanPrevOutputs()
+
+	// if challenge mode on, subscribe L2 output submission events
+	if !c.cfg.ChallengerDisabled {
+		c.wg.Add(1)
+		go c.subscribeL2OutputSubmitted()
+	}
+
+	// subscribe challenge creation events
+	c.wg.Add(1)
+	go c.subscribeChallengeCreated()
```

Comment:
Passing the current L2 block number as an explicit argument to functions like `scanPrevOutputs`, `subscribeL2OutputSubmitted`, and `subscribeChallengeCreated` can be a good approach. This allows for more flexibility and ensures that the functions are working with the correct and up-to-date L2 block number.